src/ui/connect/connector.ts

```diff
@@ -1,107 +1,133 @@
+                    reject(error);
+                } else {
+                    resolve(data);
+                }
+            });
+        });
     }
 
-    private async sendRequest<T>(request: Message, executor: (response: Message, resolve: (data?: T) => void, reject: (error: Error) => void) => void) {
-        const id = this.getRequestId();
+    private async firefoxSendRequestWithResponse<T>(name: string) {
         return new Promise<T>((resolve, reject) => {
-            const listener = ({id: responseId, ...response}: Message) => {
-                if (responseId === id) {
-                    executor(response, resolve, reject);
-                    this.port.onMessage.removeListener(listener);
+            const dataPort = chrome.runtime.connect({name});
+            dataPort.onDisconnect.addListener(() => reject());
+            dataPort.onMessage.addListener(({data, error}) => {
+                if (error) {
+                    reject(error);
+                } else {
+                    resolve(data);
                 }
-            };
-            this.port.onMessage.addListener(listener);
-            this.port.postMessage({...request, id});
+                dataPort.disconnect();
+            });
         });
     }
 
     async getData() {
-        return await this.sendRequest<ExtensionData>({type: 'get-data'}, ({data}, resolve) => resolve(data));
+        if (isFirefox) {
+            return this.firefoxSendRequestWithResponse<ExtensionData>('ui-get-data');
+        }
+        return await this.sendRequest<ExtensionData>({type: 'get-data'});
     }
 
     async getActiveTabInfo() {
-        return await this.sendRequest<TabInfo>({type: 'get-active-tab-info'}, ({data}, resolve) => resolve(data));
+        if (isFirefox) {
+            return this.firefoxSendRequestWithResponse<TabInfo>('ui-get-active-tab-info');
+        }
+        return await this.sendRequest<TabInfo>({type: 'get-active-tab-info'});
     }
 
+    private onChangesReceived = ({name, type, data}: Message) => {
+        if (name === 'background' && type === 'changes') {
+            this.changeSubscribers.forEach((callback) => callback(data));
+        }
+    };
+
     subscribeToChanges(callback: (data: ExtensionData) => void) {
-        const id = this.getRequestId();
-        this.port.onMessage.addListener(({id: responseId, data}: Message) => {
-            if (responseId === id) {
-                callback(data);
+        this.changeSubscribers.add(callback);
+        if (this.changeSubscribers.size === 1) {
+            chrome.runtime.onMessage.addListener(this.onChangesReceived);
+            if (isChromium) {
+                chrome.runtime.sendMessage({name: 'ui', type: 'subscribe-to-changes'});
             }
-        });
-        this.port.postMessage({type: 'subscribe-to-changes', id});
+        }
     }
 
     enable() {
-        this.port.postMessage({type: 'enable'});
+        chrome.runtime.sendMessage({name: 'ui', type: 'enable'});
```

Comment:
Let's move `chrome.runtime.sendMessage({name: 'ui', ...` to some method, e.g. `sendMessage(type: string, data?: any)`